conformance/packages/conformance-tests/src/name_server/rfc5155.rs

```diff
@@ -0,0 +1,376 @@
+/// name.
+fn find_cover<'a>(name_hash: &str, records: &'a BTreeMap<String, NSEC3>) -> Option<&'a NSEC3> {
+    let (hash, candidate) = records
+        // Find the greater hash that is less or equal than the name's hash.
+        .range(..=name_hash.to_owned())
+        .last()
+        // If no value is less or equal than the name's hash, it means that the name's hash is out
+        // of range and the last record covers it.
+        .or_else(|| records.last_key_value())?;
+
+    // If the found hash is exactly the name's hash, return None as it wouldn't be proving its
+    // nonexistence. Otherwise return the RR with that hash.
+    (hash != name_hash).then_some(candidate)
+}
+
+/// This proof consists of (up to) two different NSEC3 RRs:
+/// - An NSEC3 RR that matches the closest (provable) encloser.
+/// - An NSEC3 RR that covers the "next closer" name to the closest encloser.
+fn closest_encloser_proof<'a>(
+    closest_encloser_hash: &str,
+    next_closer_name_hash: &str,
+    records: &'a BTreeMap<String, NSEC3>,
+) -> Option<(&'a NSEC3, &'a NSEC3)> {
+    Some((
+        find_match(closest_encloser_hash, records)?,
+        find_cover(next_closer_name_hash, records)?,
+    ))
+}
+
+fn query_nameserver(
+    records: impl IntoIterator<Item = Record>,
+    qname: &FQDN,
+    qtype: RecordType,
+) -> Result<(BTreeMap<String, NSEC3>, DigStatus, Vec<NSEC3>)> {
+    let network = Network::new()?;
+    let mut ns = NameServer::new(&dns_test::SUBJECT, FQDN::ROOT, &network)?;
+
+    for record in records {
+        ns.add(record);
+    }
+
+    let ns = ns.sign()?;
+    // Extract the NSEC3 RRs from the signed zonefile and sort them by the hash embedded in the
+    // last label of each record's owner.
+    let nsec3_rrs = ns
+        .signed_zone_file()
+        .records
+        .iter()
+        .cloned()
+        .filter_map(|rr| {
+            let mut nsec3_rr = rr.try_into_nsec3().ok()?;
+            nsec3_rr.next_hashed_owner_name = nsec3_rr.next_hashed_owner_name.to_uppercase();
+            Some((nsec3_rr.fqdn.last_label().to_uppercase(), nsec3_rr))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    let ns = ns.start()?;
+
+    let client = Client::new(&network)?;
+    let output = client.dig(
+        *DigSettings::default().dnssec().authentic_data(),
+        ns.ipv4_addr(),
+        qtype,
+        qname,
+    )?;
+
+    let nsec3_rrs_response = output
+        .authority
+        .into_iter()
+        .filter_map(|rr| rr.try_into_nsec3().ok())
+        .collect::<Vec<_>>();
+
+    Ok((nsec3_rrs, output.status, nsec3_rrs_response))
+}
+
+#[track_caller]
+fn find_records<'a>(
+    records: &[NSEC3],
+    records_and_err_msgs: impl IntoIterator<Item = (&'a NSEC3, &'a str)>,
+) {
+    for (record, err_msg) in records_and_err_msgs {
+        records.iter().find(|&rr| rr == record).expect(err_msg);
+    }
+}
+
+const ALICE_FQDN: &str = "alice.com.";
```

Comment:
I think the names could improve here. `ALICE_FQDN` is the "TLD domain name" (2 labels) and the rest are its subdomains (3 labels). `BOB_FQDN` exists and `CHARLIE_FQDN` does not so perhaps: `TLD_FQDN` (alice), `NON_EXISTENT_FQDN` (bob) and `EXISTENT_FQDN` (charlie)?

(yes, longer to type. but readability should be prioritized)

Reply:
Yeah it makes sense to rename those, I removed bob as it wasn't really necessary